Userland/Games/2048/Game.cpp

```diff
@@ -207,6 +218,56 @@ Game::MoveOutcome Game::attempt_move(Direction direction)
     return MoveOutcome::InvalidMove;
 }
 
+void Game::add_evil_tile()
+{
+    auto worst_row = 0;
+    auto worst_column = 0;
+    auto worst_value = 2;
+
+    auto most_free_cells = m_grid_size * m_grid_size; // a big value to min() against
+    auto stalled = false;
+
+    for (size_t row = 0; row < m_grid_size && !stalled; row++)
+        for (size_t column = 0; column < m_grid_size && !stalled; column++) {
+            if (m_board[row][column] != 0)
+                continue;
+
+            for (auto value : Array { 2, 4 }) {
+                auto saved_state = *this;
+                saved_state.m_board[row][column] = value;
+
+                if (is_stalled(saved_state.m_board)) {
+                    stalled = true;
+
+                    worst_row = row;
+                    worst_column = column;
+                    worst_value = value;
+
+                    break;
+                }
+
+                auto best_outcome = 0ul;
+                for (auto direction : Array { Direction::Down, Direction::Left, Direction::Right, Direction::Up }) {
+                    auto moved_state = saved_state;
+                    auto moved = moved_state.slide_tiles(direction);
+                    if (!moved) // invalid move
+                        continue;
+                    best_outcome = max(best_outcome, get_number_of_free_cells(moved_state.board()));
```

Comment:
Functionality suggestion: you could perhaps factor the score into the comparison here? If a move leaves the player with some less space, but gives them _a lot_ more fake game points, isn't it a bit too generous for the Evil AI?

Reply:
I thought about it, but ultimately the goal is to make the player lose before they get to the target tile. Score could be used as a tie-breaker though.

Reply:
Added. Now the tile generator takes score into account as the secondary variable to minimize.